### ISSUE TITLE
Implement Multivariate Normal

### ISSUE THREAD

**Issue by tbrx** [probtorch/pytorch#1]
I'd like to add an implementation of a multivariate normal distribution with non-diagonal covariance (unless I'm mistaken, we don't yet have one of these in either the `probtorch` repo or in Pyro).

The main question is how we would like to parameterize it. I don't believe there is a multivariate normal in `tf.distributions`, rather just in `tf.contrib.distributions` (at least, according to the r1.4 API docs). The two most useful parameterizations are likely:

* mean + full covariance [MultivariateNormalFullCovariance](https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/MultivariateNormalFullCovariance)
* mean + lower triangular (e.g. Cholesky decomposition) root of covariance [MultivariateNormalTriL](https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/MultivariateNormalTriL)

but also potentially interesting is the low-rank + diagonal option [MultivariateNormalDiagPlusLowRank](https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/MultivariateNormalDiagPlusLowRank). As a first pass, though, I will ignore this, and require passing either a full covariance matrix or a lower-triangular decomposition. Looking at the design doc, do we have a consensus on how to handle different parameterizations for the same distribution?

For all of these, for a D-dimensional distribution they permit passing `D x D` e.g. covariance matrices, or `batch_size x D x D` tensors for when the covariance varies across the batch. I believe this should work for us as well.

**Comment by alicanb** [comment]
I think we went with single function, multiple params in the meeting with Pyro guys (like we have in probtorch right now)

**Comment by jwvdm** [comment]
I would say the multivariate normal and univariate normal are different distributions, in the sense that they have different return types, rather than different parameterizations of the same distribution.

**Comment by tbrx** [comment]
The question is more regarding how to handle multiple possible parameterizations of a multivariate normal. I agree the univariate normal is different.

**Comment by alicanb** [comment]
Yes, what I meant by having a single `MultivariateNormal` that encapsualtes `MultivariateNormalFullCovariance` and `MultivariateNormalTriL`. Whether to have a single parameter called `cov` and check whether it's lower triangular or having a `TriL` parameter is up to @tbrx

**Comment by fritzo** [comment]
Awesome, we're looking forward to using MVN for GPs! In case it might help, we had early PyTorch implementations of [MVN with full covariance](https://github.com/uber/pyro/blob/43973bd24eaf55b2fd816010f3ea97a0535b1385/pyro/distributions/normal.py)
 and [MVN with Cholesky covariance](https://github.com/uber/pyro/blob/43973bd24eaf55b2fd816010f3ea97a0535b1385/pyro/distributions/normal_chol.py) in pre-release Pyro. We removed these prior to launch because they had no tests and no examples.

**Comment by neerajprad** [comment]
We do not have autograd support for `potrf` and `trtrs` (which the implementations may require) currently in PyTorch, and this will be needed if we would like to take the gradients of the parameter values. See https://github.com/pytorch/pytorch/issues/440.

**Comment by alicanb** [comment]
I think all of us are here on this thread so I'm writing this here: our fork has a `distribution-api` branch I started ~2 weeks ago. You may want to branch off/push to that.

**Comment by tbrx** [comment]
@alicanb I saw that branch, and will follow your lead on formatting and structure.

@neerajprad Since pytorch#2386, `potrf` actually does have a backward implementation, so this will at least be doable. It appears that `trtrs` does not even have a CUDA implementation for `forward` at the moment, but this will be addressed with pytorch#3073. In the meantime we can fall back to either less-efficient solvers, or copying to the CPU.

**Comment by martinjankowiak** [comment]
@tbrx will it be in the scope of this issue to implement a (cuda) batched cholesky?

**Comment by tbrx** [comment]
It looks like it, unfortunately — I don't see a way around it at the moment.

We can always e.g. split along the batch dimension and then stack the results. I suppose this will be slow-ish if it is in a python list comprehension. 

On the other hand, this will only be necessary for cases in which the user passes in a batch of covariance matrices. We can keep behaviour efficient even without a batched `potrf` for users who instead parameterize the distribution with a batch of lower-triangular matrices. This will be necessary to keep runtime quadratic in number of dimensions anyway — passing in a batch of covariance matrices will always be "slow", even with a batched Cholesky. (Obviously, passing in a single covariance matrix argument will also be fine.)

**Comment by martinjankowiak** [comment]
i'd like to have batched cholesky for other reasons so i'm happy to help where i can. i'm probably not the right person to be writing cuda code but at the very least i can help with careful code review, tests, etc.

**Comment by fritzo** [comment]
@tbrx any progress on this? @dwd31415 has submitted a MultivariateNormal to Pyro https://github.com/uber/pyro/pull/651/files. Let me know if I can help in any way, as I'd like to ensure this makes it into the PyTorch 0.4 release.

**Comment by tbrx** [comment]
Just took a look at the Pyro PR — one thing that looks like it is fairly different is that the Pyro version seems to require that the `loc=…` and `covariance_matrix=…` arguments have to be vectors and matrices, respectively. Here we additionally would potentially have a `batch_shape` as well, corresponding to leading dimensions for these distribution constructor arguments.

**Comment by fritzo** [comment]
@tbrx Yeah the Pyro distributions are less general than our new PyTorch distributions. That makes it easier to implement a MultivariateNormal (no need for batched Cholesky, e.g.). What do you think about breaking this issue into two steps (two pull requests)?
1. First implement a not-very-general `torch.distributions.MultivariateNormal` that assumes vectors and matrices. When given higher order tensors this can either raise a `NotImplementedError` or implement an inefficient loop over the batched dimension. This can basically follow @dwd31415's implementation.
2. Later implement a fully-general version that uses batched linear algebra. Maybe we can work with PyTorch devs to implement CUDA versions of these things after 0.4 is released.

**Comment by tbrx** [comment]
Yeah, that is a good idea. The only real issue then is the lack of a differentiable `potri`, and we can use a workaround like in @dwd31415's code.

Allowing a batch shape for `loc`, at least, is easy enough. I'll put together a PR which throws a `NotImplementedError` when there is a batch shape on the covariance, and then we can think about whether to add initial support using python loops.

**Comment by tbrx** [comment]
Does anyone know of a currently-supported way to batch solve or batch factor matrices? Neither `torch.potrf` or `torch.gesv` have batched versions at the moment as far as I can tell. There is a batched LU decomposition in `torch.btrifact`, but unfortunately that does not currently support `.backward()`.

**Comment by fritzo** [comment]
@tbrx Since we already know the LU decomposition (`L = U' = sigma_tril`), can we `btrisolve` by reconstructing fake outputs from `torch.btrifact`? It might take a little reverse engineering to figure out the `btrifact` output format.

**Comment by fritzo** [comment]
Looks like we could use @zou3519's https://github.com/pytorch/pytorch/pull/4502

**Comment by tbrx** [comment]
Yes, that looks good! I'll implement as soon as `bgesv` is merged into upstream master.
